### PR TITLE
improvements to paqa_devs.c

### PR DESCRIPTION
**Pull request by philburk**
Improved test macros.
Test blocking read/write.
Test non-interleaved.

### CODE REVIEW

**Comment by philburk** [comment]
Note that this test crashes on Ubuntu 20 for blocking, non-interleaved streams. I will file a bug with more details.

File: qa/paqa_macros.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
+
+#ifndef PORTAUDIO_QA_PAQA_TOOLS_H
+#define PORTAUDIO_QA_PAQA_TOOLS_H
```
Comment (RossBencina) [inline qa/paqa_macros.h:3]:
```suggestion
#ifndef PORTAUDIO_QA_PAQA_MACROS_H
#define PORTAUDIO_QA_PAQA_MACROS_H
```

File: qa/paqa_macros.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        if ((_exp)) {\
+            gNumPassed++; \
+        } \
+        else { \
+            printf("\nERROR - 0x%x - %s for %s\n", result, Pa_GetErrorText(result), #_exp ); \
+            gNumFailed++; \
+        } \
+    } while(0)
+
+#endif /* PORTAUDIO_QA_PAQA_TOOLS_H */
```
Comment (RossBencina) [inline qa/paqa_macros.h:74]:
```suggestion
#endif /* PORTAUDIO_QA_PAQA_MACROS_H*/
```

File: qa/paqa_macros.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
+
+#ifndef PORTAUDIO_QA_PAQA_TOOLS_H
+#define PORTAUDIO_QA_PAQA_TOOLS_H
+
+static int gNumPassed = 0; /* Two globals */
+static int gNumFailed = 0;
```
Comment (RossBencina) [inline qa/paqa_macros.h:6]: These really are "globals". Declaring them static in a header file seems risky to me -- what happens if there is more than one `.c` file in the test program? 

I would prefer to make them real globals: declare them here, and define a macro that can be used to instantiate them in the compilation unit that contains main. E.g.

```
In this .h:

extern int gNumPassed;
extern int gNumFailed;

/* you must use this macro exactly once in each test program */
#define PAQA_INSTANTIATE_GLOBALS\
    int gNumPassed = 0;\
    int gNumPassed = 0;

In the .c file containing `main()`:

PAQA_INSTANTIATE_GLOBALS
```

maybe also rename `gpaqaNumPassed` or something more consistent with our naming scheme.

File: qa/paqa_macros.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+//    { \
+//        if ((_exp)) {\
+//            gNumPassed++; \
+//        } \
+//        else { \
+//            printf("\nERROR - 0x%x - %s for %s\n", result, Pa_GetErrorText(result), #_exp ); \
+//            gNumFailed++; \
+//            goto error; \
+//        } \
+//    } while(0)
```
Comment (RossBencina) [inline qa/paqa_macros.h:23]: Do we need to keep this commented out version?

```suggestion
```

File: qa/paqa_errs.c [hunk 5]
```diff
@@ -393,11 +363,13 @@ int main(void)
...
+    printf("paqa_errs - PortAudio QA test\n");
+    ASSERT_EQ(paNoError, (result = Pa_Initialize()));
     TestBadOpens();
     TestBadActions();
 error:
     Pa_Terminate();
-    printf("QA Report: %d passed, %d failed.\n", gNumPassed, gNumFailed);
-    return 0;
+    printf("paqa_errs: %d passed, %d failed.\n", gNumPassed, gNumFailed);
+    return (gNumFailed > 0) || (gNumPassed == 0);
```
Comment (RossBencina) [inline qa/paqa_errs.c:374]:
```suggestion
    return (gNumFailed > 0) || (gNumPassed == 0)
        ? EXIT_FAILURE
        : EXIT_SUCCESS;
```

File: qa/paqa_errs.c [hunk 6]
```diff
@@ -47,6 +47,7 @@
 #include <math.h>
```
Comment (RossBencina) [inline qa/paqa_errs.c:47]:
```suggestion
#include <stdlib.h> /* for EXIT_SUCCESS and EXIT_FAILURE */
#include <math.h>
```

File: qa/paqa_devs.c [hunk 7]
```diff
@@ -207,105 +301,368 @@ static int QaCallback( const void *inputBuffer, void *outputBuffer,
...
+    PaError result = paNoError;
+    double elapsedTime = 0.0;
+    double millisPerBurst = MILLIS_PER_SECOND * data->framesPerBurst / data->parameters->sampleRate;
+    while (elapsedTime < millis) {
+        int callbackResult;
+        if (data->parameters->mode == MODE_OUTPUT) {
+            callbackResult = QaCallback(NULL /*inputBuffer */,
+                                        data->audioBuffer,
+                                        data->framesPerBurst,
+                                        NULL, // TODO
```
Comment (RossBencina) [inline qa/paqa_devs.c:320]:
```suggestion
                                        NULL /*timeInfo */, // TODO
```

File: qa/paqa_devs.c [hunk 8]
```diff
@@ -52,83 +52,165 @@
...
+    myData->framesPerBurst = (parameters->framesPerBuffer == 0) ? 128 : parameters->framesPerBuffer;
+    if (parameters->useCallback == 0) {
+        /* We need our own buffer for blocking IO. */
+        int numChannels = (parameters->mode == MODE_OUTPUT)
+                ? parameters->numOutputChannels
+                : parameters->numInputChannels;
+        myData->audioBuffer = malloc(myData->bytesPerSample * numChannels * myData->framesPerBurst);
+    }
+}
+// TODO teardown delete audioBuffer
```
Comment (RossBencina) [inline qa/paqa_devs.c:182]: I notice this is not done. I see why you haven't done it.

File: qa/paqa_devs.c [hunk 9]
```diff
@@ -52,83 +52,165 @@
...
-        } \
-    } while(0)
-
-static float NextSineSample( PaQaData *data )
+
+static float NextSineSample( PaSineOscillator *sineOscillator )
 {
-    float phase = data->phase + PHASE_INCREMENT;
+    float phase = sineOscillator->phase + sineOscillator->phaseIncrement;
     if( phase > M_PI ) phase -= 2.0 * M_PI;
```
Comment (RossBencina) [inline qa/paqa_devs.c:189]:
```suggestion
    if( phase > M_PI ) phase -= (float)(2.0 * M_PI);
```

File: qa/paqa_devs.c [hunk 10]
```diff
@@ -317,139 +674,132 @@ static void TestDevices( int mode, int allDevices )
...
-            printf("===========================================================\n");
+            printf("\n---------------------- NumChannels = %d ------------\n", jc );
+            if (mode == MODE_INPUT) {
+                parameters.numInputChannels = jc;
+            } else {
+                parameters.numOutputChannels = jc;
+            }
             /* Try each standard sample rate. */
-            for( i=0; standardSampleRates[i] > 0; i++ )
+            for( i=0; kStandardSampleRates[i] > 0; i++ )
```
Comment (RossBencina) [inline qa/paqa_devs.c:693]: I am not comfortable with letting this old Macintosh/Android naming convention into the tree. Let's stick to our style guide.

File: qa/paqa_devs.c [hunk 11]
```diff
@@ -207,105 +301,368 @@ static int QaCallback( const void *inputBuffer, void *outputBuffer,
     {
         PaUtil_AddTraceMessage("QaCallback: DONE! framesLeft", data->framesLeft );
         data->framesLeft = 0;
-        return 1;
+        return 1; /* STOP */
```
Comment (RossBencina) [inline qa/paqa_devs.c:304]:
```suggestion
        return paComplete;
```

File: qa/paqa_errs.c [hunk 12]
```diff
@@ -63,37 +64,6 @@ typedef struct PaQaData
 }
 PaQaData;
```
Comment (RossBencina) [inline qa/paqa_errs.c:65]: looks like `framesLeft` is unused.

**Review by RossBencina — COMMENTED** [review]
Hi Phil,

I've done a first-pass over this but I have not reviewed the code with a fine toothed comb. It's a lot of code!

If you're comfortable with it I'm happy for you to merge this once you've responded to this round of review comments, but if you'd prefer me to do a more detailed review then I can do so. Just let me know.

File: qa/paqa_macros.h (hunk 4 above)
Comment (philburk) [reply qa/paqa_macros.h:23]: No

File: qa/paqa_devs.c (hunk 8 above)
Comment (philburk) [reply qa/paqa_devs.c:182]: Did it.

File: qa/paqa_devs.c (hunk 10 above)
Comment (philburk) [reply qa/paqa_devs.c:693]: I find the Hungarian notation to be very helpful and improves readability. But that is a discussion for another day. Done.

File: qa/paqa_errs.c (hunk 12 above)
Comment (philburk) [reply qa/paqa_errs.c:65]: It's used a lot.

**Pull request merged**